### PR TITLE
 [LWG motion 12 2024-06] P2300R10 std::execution

### LINKED ISSUE #7095: [2024-06 LWG Motion 12] P2300R10 std::execution
https://wiki.edg.com/pub/Wg21stlouis2024/StrawPolls/P2300R10.html

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #7095
Fixes cplusplus/papers#1054

### CODE REVIEW

**Comment by ericniebler** [comment]
fyi, there is another unescaped `<` in [exec.with.awaitable.senders] p2. See cplusplus/sender-receiver#272.

File: source/exec.tex [hunk 1]
```diff
@@ -0,0 +1,4872 @@
...
+  template<template<class...> class T, class... Args>
+  concept @\defexposconcept{valid-specialization}@ =                                // \expos
+    requires { typename T<Args...>; };
+
+  struct @\exposid{default-impls}@ {          // \expos
+    static constexpr auto @\exposid{get-attrs}@ = @\seebelow@;
+    static constexpr auto @\exposid{get-env}@ = @\seebelow@;
+    static constexpr auto @\exposid{get-state}@ = @\seebelow@;
+    static constexpr auto @\exposid{start}@ = @\seebelow@;
+    static constexpr auto @\exposid{complete}@ = @\seebelow@;
```
Comment (ericniebler) [inline source/exec.tex:1530]:
```suggestion
    static constexpr auto @\exposid{get-attrs}@ = @\seebelow@; // \expos
    static constexpr auto @\exposid{get-env}@ = @\seebelow@; // \expos
    static constexpr auto @\exposid{get-state}@ = @\seebelow@; // \expos
    static constexpr auto @\exposid{start}@ = @\seebelow@; // \expos
    static constexpr auto @\exposid{complete}@ = @\seebelow@; // \expos
```

File: source/exec.tex [hunk 2]
```diff
@@ -0,0 +1,4872 @@
...
+For \tcode{just}, \tcode{just_error}, and \tcode{just_stopped},
+let \exposid{set-cpo} be
+\tcode{set_value}, \tcode{set_error}, and \tcode{set_stopped}, respectively.
+The exposition-only class template \exposid{impls-for}\iref{exec.snd.general}
+is specialized for \exposid{just-cpo} as follows:
+\begin{codeblock}
+namespace std::execution {
+  template<>
+  struct @\exposid{impls-for<}@@\exposid{decayed-typeof}@<@\exposid{just-cpo}@>> : @\exposid{default-impls}@ {
+    static constexpr auto start =
```
Comment (ericniebler) [inline source/exec.tex:2530]:
```suggestion
  struct @\exposid{impls-for}@<@\exposid{decayed-typeof}@<@\exposid{just-cpo}@>> : @\exposid{default-impls}@ {
    static constexpr auto @\exposid{start}@ =
```

File: source/exec.tex [hunk 3]
```diff
@@ -0,0 +1,4872 @@
...
+except that \tcode{sndr} is evaluated only once.
+
+\pnum
+The exposition-only class template \exposid{impls-for}
+is specialized for \tcode{continues_on_t} as follows:
+\begin{codeblock}
+namespace std::execution {
+  template<>
+  struct @\exposid{impls-for}@<continues_on_t> : @\exposid{default-impls}@ {
+    static constexpr auto get_attrs =
```
Comment (ericniebler) [inline source/exec.tex:2795]:
```suggestion
    static constexpr auto @\exposid{get-attrs}@ =
```

File: source/exec.tex [hunk 4]
```diff
@@ -0,0 +1,4872 @@
...
+    using stop_callback = stop_callback_of_t<stop_token_of_t<env_of_t<Rcvr>>, @\exposid{on-stop-request}@>;
+
+    struct @\exposid{state-type}@ {
+      void arrive(Rcvr& rcvr) noexcept {
+        if (0 == --count) {
+          @\exposid{complete}@(rcvr);
+        }
+      }
+
+      void @\exposid{complete}@(Rcvr& rcvr) noexcept;   // \seebelow
```
Comment (ericniebler) [inline source/exec.tex:4071]:
```suggestion
      void @\exposid{arrive}@(Rcvr& rcvr) noexcept {  // \expos
        if (0 == --count) {
          @\exposid{complete}@(rcvr);
        }
      }

      void @\exposid{complete}@(Rcvr& rcvr) noexcept;   // \seebelow  // \expos
```

File: source/utilities.tex [hunk 5]
```diff
@@ -10821,6 +10821,23 @@
...
+    concept @\defexposconcept{nothrow-callable}@ =   // \expos
+      \exposconcept{callable}<Fn, Args...> &&
+      requires (Fn&& fn, Args&&... args) {
+        { std::forward<Fn>(fn)(std::forward<Args>(args)...) } noexcept;
+      };
+  template<class Fn, class... Args>
+    using @\exposid{call-result-t}@ = decltype(declval<Fn>()(declval<Args>()...));  // \expos
+
+  template<const auto& T>
+    using @\exposid{decayed-typeof}@ = decltype(auto(T));                           // \expos
```
Comment (frederick-vs-ja) [inline source/utilities.tex:10840]: Can we just reuse the existing `decay_t`? (Perhaps changes should be in another PR, although.)

**Comment by jensmaurer** [comment]
We should clearly differentiate between fixes where the transcription from the approved P2300R10 HTML paper is wrong vs. fixes that go on top of that paper, because the paper is wrong / incomplete / broken / whatever.

The former fixes will eventually be squashed; the latter ones should remain separate commits to allow for easier historic inspection.

File: source/exec.tex (hunk 1 above)
Comment (jensmaurer) [reply source/exec.tex:1577]: Fixed as "not in the incoming paper".

File: source/exec.tex (hunk 3 above)
Comment (jensmaurer) [reply source/exec.tex:2842]: Same problem for [exec.run.loop.general] push-back and pop-front 
Both issues are present in the incoming paper.

File: source/exec.tex (hunk 2 above)
Comment (jensmaurer) [reply source/exec.tex:2577]: The first issue is fixed (obviously a transcription problem).
The second issue is also fixed, but was already present in the incoming paper.

**Comment by ericniebler** [comment]
hi @jensmaurer. what remains to be done on this pr?

File: source/exec.tex [hunk 6]
```diff
@@ -0,0 +1,5466 @@
...
+\begin{itemize}
+\item
+is explicitly created;
+\item
+can be explicitly started once at most;
+\item
+once started, eventually completes exactly once
+with a (possibly empty) set of result datums and
+in exactly one of three \defnx{disposition}{dispositions}:
+success, failure, or cancellation;
```
Comment (ericniebler) [inline source/exec.tex:190]: in the pdf, this is rendered as "_disposition_" (singlular) should be plural.
image

**Comment by jensmaurer** [comment]
> hi @jensmaurer. what remains to be done on this pr?

Add indexing, merge to "main" branch.

**Comment by tkoeppe** [comment]
I have rebased this onto the current state; any reviews welcome!

File: source/exec.tex (hunk 6 above)
Comment (tkoeppe) [reply source/exec.tex:190]: Fixing, thanks!
